tests/dbshell/test_postgresql_psycopg2.py

```diff
@@ -0,0 +1,106 @@
+# -*- coding: utf8 -*-
+import locale
+import os
+
+from django.db.backends.postgresql_psycopg2.client import DatabaseClient
+from django.test import SimpleTestCase, mock
+from django.utils.encoding import force_bytes, force_str
+
+
+class PostgreSqlDbshellCommandTestCase(SimpleTestCase):
+
+    def _run_it(self, dbinfo):
+        '''
+        That function invokes the runshell command, while mocking
+        subprocess.call. It returns a 2-tuple with:
+        - The command line list
+        - The binary content of file pointed by environment PGPASSFILE, or
+          None.
+        '''
+        def _mock_subprocess_call(*args):
+            self.subprocess_args = list(*args)
+            if 'PGPASSFILE' in os.environ:
+                self.pgpass = open(os.environ['PGPASSFILE'], 'rb').read()
+            else:
+                self.pgpass = None
+            return 0
+        self.subprocess_args = None
+        self.pgpass = None
+        with mock.patch('subprocess.call', new=_mock_subprocess_call):
+            DatabaseClient.runshell_db(dbinfo)
+        return self.subprocess_args, self.pgpass
+
+    def test_basic(self):
+        self.assertEqual(
+            self._run_it({
+                'NAME': 'somedbname',
+                'USER': 'someuser',
+                'PASSWORD': 'somepassword',
+                'HOST': 'somehost',
+                'PORT': 444,
+            }), (
+                ['psql', '-U', 'someuser', '-h', 'somehost', '-p', '444',
+                 'somedbname'],
+                b'somehost:444:somedbname:someuser:somepassword\n'))
+
+    def test_nopass(self):
+        self.assertEqual(
+            self._run_it({
+                'NAME': 'somedbname',
+                'USER': 'someuser',
+                'HOST': 'somehost',
+                'PORT': 444,
+            }), (
+                ['psql', '-U', 'someuser', '-h', 'somehost', '-p', '444',
+                 'somedbname'],
+                None))
+
+    def test_column(self):
+        self.assertEqual(
+            self._run_it({
+                'NAME': 'somedbname',
+                'USER': 'some:user',
+                'PASSWORD': 'some:password',
+                'HOST': '::1',
+                'PORT': 444,
+            }), (
+                ['psql', '-U', 'some:user', '-h', '::1', '-p', '444',
+                 'somedbname'],
+                b'\\:\\:1:444:somedbname:some\\:user:some\\:password\n'))
+
+    def test_escape(self):
+        self.assertEqual(
+            self._run_it({
+                'NAME': 'somedbname',
+                'USER': 'some\\user',
+                'PASSWORD': 'some\\password',
+                'HOST': 'somehost',
+                'PORT': 444,
+            }), (
+                ['psql', '-U', 'some\\user', '-h', 'somehost', '-p', '444',
+                 'somedbname'],
+                b'somehost:444:somedbname:some\\\\user:some\\\\password\n'))
+
+    def test_accent(self):
+        encoding = locale.getpreferredencoding()
```

Comment:
I tried installing with a virtual Debian machine, I did a "dpkg-reconfigure locales" and I selected locale fr_FR whose default encoding is ISO-8859-1. This is quite unusual since most people use fr_FR.UTF-8 nowadays. I did that before installing postgres.
The pgpass temporary file needs to be encoded using system locale.
"rôle" must be encoded using 5 bytes on utf-8 systems, it must be encoded using 4 bytes on iso-8859-1 systems.